maitake-sync/src/blocking/mutex.rs

```diff
@@ -0,0 +1,465 @@
+use crate::{
+    blocking::DefaultMutex,
+    loom::cell::{MutPtr, UnsafeCell},
+    util::fmt,
+};
+use core::ops::{Deref, DerefMut};
+
+pub use mutex_traits::{RawMutex, ScopedRawMutex};
+
+/// A blocking mutual exclusion lock for protecting shared data.
+/// Each mutex has a type parameter which represents
+/// the data that it is protecting. The data can only be accessed through the
+/// RAII guards returned from [`lock`] and [`try_lock`], or within the closures
+/// passed to [`with_lock`] and [`try_with_lock`], which guarantees that
+/// the data is only ever accessed when the mutex is locked.
+///
+/// # Fairness
+///
+/// This is *not* a fair mutex.
+///
+/// # Overriding mutex implementations
+///
+/// This type is generic over a `Lock` type parameter which represents a raw
+/// mutex implementation. By default, this is the [`DefaultMutex`]. To construct
+/// a new `Mutex` with an alternative raw mutex implementation, use the
+/// [`Mutex::with_raw_mutex`] cosntructor. See the [module-level documentation
+/// on overriding mutex
+/// implementations](crate::blocking#overriding-mutex-implementations) for
+/// more details.
+///
+/// When `Lock` implements the [`RawMutex`] trait, the [`Mutex`] type provides
+/// the [`lock`] and [`try_lock`] methods, which return a RAII [`MutexGuard`],
+/// similar to the [`std::sync::Mutex`] API, in addition to the scoped
+/// [`with_lock`] and  [`try_with_lock`] methods. When `Lock` only implements
+/// [`ScopedRawMutex`], the [`Mutex`] type provides only the scoped
+/// [`with_lock`] and  [`try_with_lock`] methods.
+///
+/// :warning: Note that [`DefaultMutex`] does *not* implement `RawMutex`, so
+/// using the [`lock`] and [`try_lock`] RAII API requires selecting an
+/// alternative [`RawMutex`] implementation.
+///
+/// # Loom-specific behavior
+///
+/// When `cfg(loom)` is enabled, this mutex will use Loom's simulated atomics,
+/// checked `UnsafeCell`, and simulated spin loop hints.
+///
+/// [`lock`]: Mutex::lock
+/// [`try_lock`]: Mutex::try_lock
+/// [`with_lock`]: Mutex::with_lock
+/// [`try_with_lock`]: Mutex::try_with_lock
+/// [`std::sync::Mutex`]: https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html
+pub struct Mutex<T, Lock = DefaultMutex> {
+    lock: Lock,
+    data: UnsafeCell<T>,
+}
+
+/// An RAII implementation of a "scoped lock" of a mutex. When this structure is
+/// dropped (falls out of scope), the lock will be unlocked.
+///
+/// The data protected by the mutex can be accessed through this guard via its
+/// [`Deref`] and [`DerefMut`] implementations.
+///
+/// This structure is created by the [`lock`] and [`try_lock`] methods on
+/// [`Mutex`].
+///
+/// [`lock`]: Mutex::lock
+/// [`try_lock`]: Mutex::try_lock
+#[must_use = "if unused, the `Mutex` will immediately unlock"]
+pub struct MutexGuard<'a, T, Lock: RawMutex> {
```

Comment:
Did you want to stick with the single pointer version of this? No opinion either way, I just know we discussed it elsewhere

Reply:
That doesn't work here, since we need to use the `loom::MutPtr` type instead of a real pointer. Possibly it could change based on the cfg, but that makes the real and test code diverge more...in any case, it can always be changed in another branch.